File: 2023/csharp/Day4/Part1.cs
using Lib;

namespace Day4;

public class Part1 : IPart
{
    public int Execute()
    {
        var cards = ReadCards();
        var endResult = 0;

        foreach (var card in cards)
        {
            var result = 0;
            foreach (var number in card.MyNumbers)
            {
                if (card.WinningNumbers.Contains(number))
                {
                    if (result == 0)
                    {
                        result = 1;
                    }
                    else
                    {
                        result *= 2;
                    }
                }
            }

            endResult += result;
        }
        return endResult;
    }

    private IEnumerable<Card> ReadCards()
    {
        var result = new List<Card>();
        using var reader = new StreamReader("./input.txt");
        while (!reader.EndOfStream)
        {
            var line = reader.ReadLine();
            if (line == null) continue;
            var numbers = line.Split(":")[1];
            var splitNumbers = numbers.Split("|");
            var winningNumbers = splitNumbers[0].Trim().Split(" ").Where(s => s != string.Empty).Select(int.Parse).ToList();
            var myNumbers = splitNumbers[1].Trim().Split(" ").Where(s => s != string.Empty).Select(int.Parse).ToList();
            result.Add(new Card()
            {
                WinningNumbers = winningNumbers,
                MyNumbers = myNumbers
            });
        }

        return result;
    }
}

File: 2023/rust/day-4/src/part1.rs
pub fn process(input: &str) -> u32 {
    input.lines().map(|line| {
        let mut split_card = line.split(":");
        let _ = split_card.next();
        let numbers = split_card.next().unwrap();
        let mut split_numbers = numbers.split("|");
        let win_numbers = split_numbers.next().unwrap();
        let my_numbers = split_numbers.next().unwrap();
        let win_numbers = win_numbers
        .split(|c| c == ' ')
        .filter_map(|number| number.trim().parse::<u8>().ok()).collect::<Vec<_>>();
        let my_numbers = my_numbers
            .split(|c| c == ' ')
            .filter_map(|number| number.trim().parse::<u8>().ok());

        let count = my_numbers.filter(|my|win_numbers.contains(my)).count();

        if count == 0 {
            0
        } else {
            let mut result = 1;

            for _ in 0..(count - 1) {
                result = result * 2;
            }
            result
        }
    }).sum()
}